### PR TITLE
[Android] Save back handling state in Activity/Fragment bundle

### PR DESCRIPTION
**Pull request by gmackall**
Fixes https://github.com/flutter/flutter/issues/159158, and fixes b/355556397

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by justinmc — APPROVED** [review]
LGTM from my side :+1: . This would have taken me quite a long time to figure out on my own so thank you!

So this is unrelated to predictive back being unsupported on FlutterFragmentActivity?

**Comment by reidbaker** [comment]
> I also think we may want to restore the ability for people to opt out, so that FlutterActivity and FlutterFragmentActivity have the same behavior wrt the `android:enableOnBackInvokedCallback` (despite this sort of running counter to what the android docs suggest) #56565 (still need to find out the best way to write a test that interacts with the xml attribute) Does that make sense to you?

For now I think we should make decisions as if we will be supporting predictive back and legacy back forever.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 1]
```diff
@@ -634,6 +634,12 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
 
     super.onCreate(savedInstanceState);
 
+    if (savedInstanceState != null
+        && savedInstanceState.getBoolean(
+            FlutterActivityAndFragmentDelegate.ON_BACK_CALLBACK_ENABLED_KEY)) {
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:639]: Are you following an existing pattern in our code? If not then I suggest having this key hold if callback is enabled. That way no key indicates "default" and a true or false indicate a preference.

**Comment by justinmc** [comment]
Thanks for the detailed explanation on what's going on here. I'm on board with allowing opt-out and supporting both predictive and non-predictive back.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 1 above)
Comment (gmackall) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:639]: Just to be clear, do you mean replacing with something like
```java
if (savedInstanceState != null) {
    Boolean frameworkHandlesBack = savedInstanceState.getBoolean(
            FlutterActivityAndFragmentDelegate.ON_BACK_CALLBACK_ENABLED_KEY));
    setFrameworkHandlesBack(frameworkHandlesBack);
}
```

Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:639]: yes

Comment (gmackall) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: done!

File: shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java [hunk 2]
```diff
@@ -94,6 +96,48 @@ public void flutterViewHasId() {
     assertTrue(activity.findViewById(FlutterActivity.FLUTTER_VIEW_ID) instanceof FlutterView);
   }
 
+  @Test
+  @Config(sdk = API_LEVELS.API_34)
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java:100]: please make this minSdk not sdk

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 3]
```diff
@@ -1071,6 +1072,12 @@ public void onAttach(@NonNull Context context) {
   @Override
   public void onCreate(@Nullable Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
+    if (savedInstanceState != null) {
+      boolean frameworkHandlesBack =
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:1076]: No need to change the pr now but when we have a second variable then all of this code should go into a "save everything" and "restore everything" methods.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java [hunk 4]
```diff
@@ -94,6 +96,48 @@ public void flutterViewHasId() {
...
+
+    flutterActivityScenario.recreate();
+    flutterActivityScenario.onActivity(activity -> assertTrue(activity.hasRegisteredBackCallback));
+
+    // Clean up.
+    flutterActivityScenario.close();
+  }
+
+  @Test
+  @Config(sdk = API_LEVELS.API_34)
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java:125]: same here

**Review by reidbaker — APPROVED** [review]

File: shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java (hunk 2 above)
Comment (gmackall) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterActivityTest.java]: Done in both places, and in `FlutterFragmentActivityTest.java`

**Pull request merged**